File: AIReaderKindle/src/Services/BookCorpus.cpp
#include "BookCorpus.hpp"

#include "EpubLoader.hpp"

BookCorpus::BookCorpus(std::vector<Book> books) : books_(std::move(books)) {}

void BookCorpus::provide(long long bookId, std::vector<std::string> chapters) {
    std::lock_guard<std::mutex> lock(mutex_);
    chapters_[bookId] = std::move(chapters);
    errors_.erase(bookId);
}

void BookCorpus::loadMissing() {
    for (const auto& book : books_) {
        if (chapters_.count(book.id) || errors_.count(book.id)) continue;
        std::string error;
        auto document = EpubLoader::load(book.path, &error, false);
        if (!document) {
            errors_[book.id] = book.title + ": " + error;
            continue;
        }
        std::vector<std::string> chapters;
        for (auto& chapter : document->chapters) chapters.push_back(std::move(chapter.text));
        chapters_[book.id] = std::move(chapters);
    }
}

std::vector<SearchHit> BookCorpus::search(const std::string& query, int limit, const std::optional<BookPosition>& upTo) {
    std::lock_guard<std::mutex> lock(mutex_);
    loadMissing();
    std::vector<SearchHit> hits;
    for (const auto& book : books_) {
        auto loaded = chapters_.find(book.id);
        if (loaded == chapters_.end()) continue;
        bool bounded = upTo && upTo->bookId == book.id;
        const std::vector<std::string>& chapters = loaded->second;
        for (size_t chapter = 0; chapter < chapters.size() && static_cast<int>(hits.size()) < limit; ++chapter) {
            if (bounded && static_cast<int>(chapter) > upTo->chapter) break;
            for (SearchHit& hit : BookSearch::find(chapters[chapter], query, static_cast<int>(chapter), limit - static_cast<int>(hits.size()))) {
                if (bounded && static_cast<int>(chapter) == upTo->chapter && hit.offset >= upTo->offset) break;
                hit.bookId = book.id;
                hit.bookTitle = book.title;
                hits.push_back(std::move(hit));
            }
        }
    }
    return hits;
}

std::vector<std::string> BookCorpus::errors() {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<std::string> list;
    for (const auto& entry : errors_) list.push_back(entry.second);
    return list;
}

File: AIReader/AIReader/Services/Books/BookCorpus.swift
import Foundation

/// The text of the book being read and of the other books in its group,
/// loaded once, when first searched, and searched from any task.
actor BookCorpus {
    /// In the order they are to be searched: the open book first.
    nonisolated let books: [Book]

    private var documents: [Book.ID: BookDocument] = [:]
    private var failures: [Book.ID: String] = [:]

    init(books: [Book]) {
        self.books = books
    }

    nonisolated var severalBooks: Bool { books.count > 1 }

    /// Hands over the book the reader has already loaded, so it is not read
    /// again.
    func provide(_ document: BookDocument, for bookID: Book.ID) {
        documents[bookID] = document
        failures[bookID] = nil
    }

    /// Hits book by book, in reading order, at most `limit`. With `upTo`,
    /// that book is searched only as far as that point — what the reader has
    /// seen — and the others in full.
    func search(_ query: String, limit: Int, upTo: BookPosition? = nil) async -> [SearchHit] {
        await loadMissing()
        var hits: [SearchHit] = []
        for book in books where hits.count < limit {
            guard let document = documents[book.id] else { continue }
            var range: NSRange?
            if let upTo, upTo.bookID == book.id {
                range = NSRange(location: 0, length: min(upTo.offset, document.text.length))
            }
            let found = BookSearch.find(
                in: document.words.text,
                query: query,
                range: range,
                limit: limit - hits.count
            )
            hits += found.map { hit in
                var hit = hit
                hit.bookID = book.id
                hit.bookTitle = book.title
                hit.chapter = document.chapter(containing: hit.offset)
                return hit
            }
        }
        return hits
    }

    /// Why a book could not be read, if one could not; checked after a search.
    var errors: [String] {
        books.compactMap { failures[$0.id] }
    }

    private func loadMissing() async {
        for book in books where documents[book.id] == nil && failures[book.id] == nil {
            do {
                documents[book.id] = try await BookDocumentLoader.load(
                    folder: book.folder,
                    packagePath: book.packagePath
                )
            } catch {
                failures[book.id] = "\(book.title): \(error.localizedDescription)"
            }
        }
    }
}

/// One corpus is the same as itself and no other, which is all a feature's
/// state needs to know.
extension BookCorpus: Equatable {
    nonisolated static func == (lhs: BookCorpus, rhs: BookCorpus) -> Bool { lhs === rhs }
}

/// What is open in front of the reader, as a lookup, an X-ray or a
/// conversation sees it: the books to search, and how far they have read.
struct ReadingScope: Equatable, Sendable {
    var corpus: BookCorpus?
    var upTo: BookPosition?

    static let none = ReadingScope()
}
